validation/url_checker.py

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Validates schema urls."""
+import csv
+import git
+import os
+import pathlib
+import re
+import requests
+
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+flags.DEFINE_boolean("update_allowlist", False,
+                     "If set, regenerates url_allowlist.csv.")
+
+API_ERROR = "An HTTP {} code was returned by the autopush API. Please ensure that DC_API_KEY is set to the autopush API key."
+API_PREFIX = "https://autopush.api.datacommons.org/v2/node?key="
+MANIFEST = "MANIFEST"
+URL_ALLOWLIST = "url_allowlist.csv"
+URL_PROPS = {"license", "url"}
+URL_REGEX = (
+    r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"
+)
+
+
+def get_status(url):
+  """Gets the status for invalid url.
+
+  Args:
+    url: Url to check.
+
+  Returns:
+    status: Status code and reason.
+    message: Optional message providing additional details about response.
+  """
+  try:
+    resp = requests.get(url, timeout=30)
+
+    # Redirect.
+    if resp.history:
+      initial_resp = resp.history[0]
+      return str(initial_resp.status_code
+                ) + " " + initial_resp.reason if initial_resp.reason else str(
+                    initial_resp.status_code), "Redirected to " + resp.url
+
+    return str(resp.status_code) + " " + resp.reason if resp.reason else str(
+        resp.status_code), ""
+
+  except Exception as e:
+    return "ERROR", str(e)
+
+
+def get_modified_files():
+  """Returns list of locally modified MCF files in the current branch.
+
+  Returns:
+    List of modified files.
+  """
+  repo = git.Repo('.', search_parent_directories=True)
+  return [
+      "../" + file for file in sorted(
+          filter(
+              lambda x: x.endswith(".mcf"),
+              list(item.a_path for item in repo.index.diff(None)) +
+              list(repo.untracked_files)))
+  ]
+
+
+def find_invalid_manifest_urls(allowlisted_urls):
+  """Finds invalid urls from manifest using the DC API.
+
+  Args:
+    allowlisted_urls: Set of urls to skip checking.
+
+  Returns:
+    Map of ("MANIFEST", <url>) -> (<status>, <message>) for invalid urls.
+  """
+  dc_api_key = os.environ.get('DC_API_KEY', '')
+
+  resp = requests.get(
+      API_PREFIX + dc_api_key +
+      "&nodes=Provenance&nodes=Dataset&nodes=Source&property=<-typeOf")
+  if resp.status_code != 200:
```

Comment:
To make sure, are there any limits at play here or should we use pagination apis?

Reply:
Good point! Added pagination